nifi-commons/nifi-record/src/test/java/org/apache/nifi/serialization/TestSimpleRecordSchema.java

```diff
@@ -93,21 +93,142 @@ public void testHashCodeAndEqualsWithSelfReferencingSchema() {
+    @Test
+    public void testNotEqualsSimpleSchemaDifferentTypes() {
+        // GIVEN
+        final String nameOfField1 = "field1";
+        final String nameOfField2 = "field2";
+        final DataType typeOfField1 = RecordFieldType.INT.getDataType();
+        final DataType typeOfField2 = RecordFieldType.STRING.getDataType();
+        final String schemaName = "schemaName";
+        final String namespace = "namespace";
+
+        final SimpleRecordSchema schema1 = createSchemaWithTwoFields(nameOfField1, nameOfField2, typeOfField1, typeOfField1, schemaName, namespace);
+        final SimpleRecordSchema schema2 = createSchemaWithTwoFields(nameOfField1, nameOfField2, typeOfField1, typeOfField2, schemaName, namespace);
+
+        // WHEN, THEN
+        assertFalse(schema1.equals(schema2));
+    }
+
+    @Test
+    public void testNotEqualsSimpleSchemaDifferentFieldNames() {
+        // GIVEN
+        final String nameOfField1 = "field1";
+        final String nameOfField2 = "field2";
+        final String nameOfField3 = "field3";
+        final DataType typeOfField1 = RecordFieldType.INT.getDataType();
+        final DataType typeOfField2 = RecordFieldType.STRING.getDataType();
+        final String schemaName = "schemaName";
+        final String namespace = "namespace";
+
+        final SimpleRecordSchema schema1 = createSchemaWithTwoFields(nameOfField1, nameOfField2, typeOfField1, typeOfField2, schemaName, namespace);
+        final SimpleRecordSchema schema2 = createSchemaWithTwoFields(nameOfField1, nameOfField3, typeOfField1, typeOfField2, schemaName, namespace);
+
+        // WHEN, THEN
+        assertFalse(schema1.equals(schema2));
+    }
+
+    @Test
+    public void testEqualsRecursiveSchema() {
+        final String field1 = "field1";
+        final String field2 = "field2";
+        final String schemaName = "schemaName";
+        final String namespace = "namespace";
+
+        final SimpleRecordSchema schema1 = createRecursiveSchema(field1, field2, schemaName, namespace);
+        final SimpleRecordSchema schema2 = createRecursiveSchema(field1, field2, schemaName, namespace);
 
         assertTrue(schema1.equals(schema2));
-        Mockito.verify(schema2, Mockito.never()).getFields();
+        assertTrue(schema2.equals(schema1));
+    }
+
+    @Test(expected = StackOverflowError.class)
+    public void testNotEqualsRecursiveSchemaIfSchemaNameIsDifferent() {
+        final String field1 = "field1";
+        final String field2 = "field2";
+        final String schemaName1 = "schemaName1";
+        final String schemaName2 = "schemaName2";
+        final String namespace = "namespace";
+
+        final SimpleRecordSchema schema1 = createRecursiveSchema(field1, field2, schemaName1, namespace);
+        final SimpleRecordSchema schema2 = createRecursiveSchema(field1, field2, schemaName2, namespace);
+
+        assertFalse(schema1.equals(schema2)); // Causes StackOverflowError
+    }
+
+    @Test(expected = StackOverflowError.class)
```

Comment:
I know that this is super weird. This was the behavior even before I started modifying the code and I want to draw attention to this by creating this test case.